blobby/blob.py

```diff
@@ -1,248 +0,0 @@
-            f"Response from S3 of type {type(response['Body'])}, "
-            f"not StreamingBody.")
-
-        return response['Body'].read()  # type: ignore
-
-    def readinto(self, b: bytearray) -> int:
-        to_read = min(len(self) - self.position, len(b))
-        b[:to_read] = self.read(to_read)
-        return to_read
-
-    def readable(self) -> bool:
-        return True
-
-    def tell(self) -> int:
-        return self.position
-
-    def __repr__(self) -> str:
-        return f'<{type(self).__name__} wrapping {self.object}>'
-
-    def __len__(self) -> int:
-        return self.object.content_length
-
-
-WriterCallback = Callable[[str], None]
-@inherit_docstrings
-class BlobWriter(io.BufferedIOBase):
-    """A file-like object wrapping an S3 blob with an automatic key.
-
-    Add contents to the internal buffer with `write`. No data is
-    transferred until `flush` is called.
-    """
-    def __init__(self, s3_bucket: S3Bucket) -> None:
-        self.bucket = s3_bucket
-        self.checksum = hashlib.sha256()
-        self.buffer = io.BytesIO()
-        self.callbacks: List[WriterCallback] = []
-
-    def register_callback(self, cb: WriterCallback) -> None:
-        """Register the callable cb to be called when the writer is flushed.
-
-        Argument:
-            cb (callable(key)): will be called with whatever key was
-                written to when the writer is flushed.
-        """
-        self.callbacks.append(cb)
-
-    def readable(self) -> bool:
-        return False
-
-    def writable(self) -> bool:
-        if self.closed:
-            raise ValueError(
-                f'I/O operation on closed {type(self).__name__}'
-            )
-
-        return True
-
-    def seekable(self) -> bool:
-        return False
-
-    def write(self, b: bytes) -> int:
-        if self.closed:
-            raise ValueError(
-                f'I/O operation on closed {type(self).__name__}'
-            )
-
-        self.checksum.update(b)
-        self.buffer.write(b)
-
-        return len(b)
-
-    def key(self) -> str:
-        return self.checksum.hexdigest()
-
-    def flush(self) -> None:
-        # FIXME: it doesn't really make sense to call this multiple
-        #   times, so this doesn't match flush semantics. S3 doesn't
-        #   support partial uploads so we'd have to use a buffer which
-        #   didn't auto-EOF but blocked until being closed. Then this
-        #   could push more data into that buffer. But there could be
-        #   problems with long-lasting HTTP connections etc.
-        if self.closed:
-            raise ValueError(f'I/O operation on closed {type(self).__name__}')
-
-        if self.buffer.closed:
-            return
-
-        nbytes = self.buffer.getbuffer().nbytes
-        key = self.key()
-        self.buffer.seek(0)
-        logger.debug('Uploading %d bytes to %s', nbytes, key)
-        # boto3 closes the buffer after this.
```

Comment:
Is this a public part of the `boto3` API or an observation? Might be worth adding in a check here.

Reply:
Just an observation. I will ensure consistency. It's a bit annoying though because it would be better to delay closing our buffer until a point of our choosing :P